--- a/src/options.c
+++ b/src/options.c
@@ -38,7 +38,11 @@
 #include "prototypes.h"

 #if OPENSSL_VERSION_NUMBER >= 0x10101000L
+#ifdef OPENSSL_NO_ECX
+#define DEFAULT_CURVES "P-256:P-521:P-384"
+#else
 #define DEFAULT_CURVES "X25519:P-256:X448:P-521:P-384"
+#endif /* OPENSSL_NO_ECX */
 #ifdef SSL_SYSTEM_DEFAULT_CIPHER_LIST /* Red Hat OpenSSL */
 #define DEFAULT_CURVES_FIPS "P-256:P-521:P-384"
 #else /* standard OpenSSL */
