Index: u-boot-2021.01/tools/fit_image.c
===================================================================
--- u-boot-2021.01.orig/tools/fit_image.c
+++ u-boot-2021.01/tools/fit_image.c
@@ -21,6 +21,8 @@
 #include <stdarg.h>
 #include <version.h>
 #include <u-boot/crc.h>
+#include <libgen.h>
+#include <ext_sign.h>
 
 static image_header_t header;
 
@@ -33,6 +35,8 @@ static int fit_add_file_data(struct imag
 	struct stat sbuf;
 	void *ptr;
 	int ret = 0;
+	char *tmp;
+	struct ext_sign_info ext_sign = {0};
 
 	tfd = mmap_fdt(params->cmdname, tmpfile, size_inc, &ptr, &sbuf, true,
 		       false);
@@ -52,6 +56,14 @@ static int fit_add_file_data(struct imag
 		destfd_size = dest_sbuf.st_size;
 	}
 
+	/* set final output image directory as temporary dir for data binaries and signature */
+	if (params->sign_tool) {
+		snprintf(ext_sign.signer_path, sizeof(ext_sign.signer_path), "%s", params->sign_tool);
+		tmp = strdup(params->imagefile);
+		snprintf(ext_sign.work_dir, sizeof(ext_sign.work_dir), "%s", dirname(tmp));
+		free(tmp);
+	}
+
 	/* for first image creation, add a timestamp at offset 0 i.e., root  */
 	if (params->datafile || params->reset_timestamp) {
 		time_t time = imagetool_get_source_date(params->cmdname,
@@ -61,10 +73,9 @@ static int fit_add_file_data(struct imag
 
 	if (!ret && params->keyname) {
 		ret = fit_add_certificate_data(ptr, params->keydir, params->require_keys,
-			 params->keyname, params->algo);
+			 params->keyname, params->algo, params->sign_tool ? &ext_sign : NULL);
 	}
 
-
 	if (!ret) {
 		ret = fit_cipher_data(params->keydir, dest_blob, ptr,
 				      params->comment,
@@ -80,7 +91,8 @@ static int fit_add_file_data(struct imag
 						params->engine_id,
 						params->cmdname,
 						params->keyname ? params->keyname : NULL,
-						params->algo ? params->algo : NULL);
+						params->algo ? params->algo : NULL,
+						params->sign_tool ? &ext_sign : NULL);
 	}
 
 	if (dest_blob) {
@@ -800,7 +812,7 @@ static int fit_handle_file(struct image_
 	 * would be considerably more complex to implement. Generally a few
 	 * steps of this loop is enough to sign with several keys.
 	 */
-	for (size_inc = 0; size_inc < 64 * 1024; size_inc += 1024) {
+	for (size_inc = 32 * 1024; size_inc < 64 * 1024; size_inc += 4 * 1024) {
 		if (copyfile(bakfile, tmpfile) < 0) {
 			printf("Can't copy %s to %s\n", bakfile, tmpfile);
 			ret = -EIO;
Index: u-boot-2021.01/tools/image-host.c
===================================================================
--- u-boot-2021.01.orig/tools/image-host.c
+++ u-boot-2021.01/tools/image-host.c
@@ -13,6 +13,7 @@
 #include <fdt_region.h>
 #include <image.h>
 #include <version.h>
+#include <ext_sign.h>
 
 /**
  * fit_set_hash_value - set hash value in requested has node
@@ -211,7 +212,7 @@ static int fit_image_process_sig(const c
 		void *fit, const char *image_name,
 		int noffset, const void *data, size_t size,
 		const char *comment, int require_keys, const char *engine_id,
-		const char *cmdname)
+		const char *cmdname, struct ext_sign_info *ext_sign)
 {
 	struct image_sign_info info;
 	struct image_region region;
@@ -227,7 +228,11 @@ static int fit_image_process_sig(const c
 	node_name = fit_get_name(fit, noffset, NULL);
 	region.data = data;
 	region.size = size;
-	ret = info.crypto->sign(&info, &region, 1, &value, &value_len);
+
+	info.ext_sign = ext_sign;
+	ret = info.crypto->sign(&info, &region, 1, &value,
+				&value_len);
+
 	if (ret) {
 		printf("Failed to sign '%s' signature node in '%s' image node: %d\n",
 		       node_name, image_name, ret);
@@ -600,7 +605,8 @@ int fit_image_cipher_data(const char *ke
  */
 int fit_image_add_verification_data(const char *keydir, void *keydest,
 		void *fit, int image_noffset, const char *comment,
-		int require_keys, const char *engine_id, const char *cmdname)
+		int require_keys, const char *engine_id, const char *cmdname,
+		struct ext_sign_info *ext_sign)
 {
 	const char *image_name;
 	const void *data;
@@ -637,7 +643,7 @@ int fit_image_add_verification_data(cons
 				strlen(FIT_SIG_NODENAME))) {
 			ret = fit_image_process_sig(keydir, keydest,
 				fit, image_name, noffset, data, size,
-				comment, require_keys, engine_id, cmdname);
+				comment, require_keys, engine_id, cmdname, ext_sign);
 		}
 		if (ret)
 			return ret;
@@ -895,7 +901,7 @@ static int fit_config_get_data(void *fit
 static int fit_config_process_sig(const char *keydir, void *keydest,
 		void *fit, const char *conf_name, int conf_noffset,
 		int noffset, const char *comment, int require_keys,
-		const char *engine_id, const char *cmdname)
+		const char *engine_id, const char *cmdname, struct ext_sign_info *ext_sign)
 {
 	struct image_sign_info info;
 	const char *node_name;
@@ -916,8 +922,10 @@ static int fit_config_process_sig(const
 				require_keys ? "conf" : NULL, engine_id))
 		return -1;
 
+	info.ext_sign = ext_sign;
 	ret = info.crypto->sign(&info, region, region_count, &value,
 				&value_len);
+
 	free(region);
 	if (ret) {
 		printf("Failed to sign '%s' signature node in '%s' conf node\n",
@@ -959,11 +967,12 @@ static int fit_config_process_sig(const
 
 static int fit_config_add_verification_data(const char *keydir, void *keydest,
 		void *fit, int conf_noffset, const char *comment,
-		int require_keys, const char *engine_id, const char *cmdname)
+		int require_keys, const char *engine_id, const char *cmdname,
+		struct ext_sign_info *ext_sign)
 {
 	const char *conf_name;
 	int noffset;
-
+	int n_sig = 0;
 	conf_name = fit_get_name(fit, conf_noffset, NULL);
 
 	/* Process all hash subnodes of the configuration node */
@@ -973,12 +982,18 @@ static int fit_config_add_verification_d
 		const char *node_name;
 		int ret = 0;
 
+		if (ext_sign) {
+			snprintf(ext_sign->data_file, sizeof(ext_sign->data_file), "%s_%d_%s", conf_name, n_sig, DATAFILE_NAME);
+			snprintf(ext_sign->sig_file, sizeof(ext_sign->sig_file), "%s_%d_%s", conf_name, n_sig, SIGFILE_NAME);
+			n_sig ++;
+		}
+
 		node_name = fit_get_name(fit, noffset, NULL);
 		if (!strncmp(node_name, FIT_SIG_NODENAME,
 			     strlen(FIT_SIG_NODENAME))) {
 			ret = fit_config_process_sig(keydir, keydest,
 				fit, conf_name, conf_noffset, noffset, comment,
-				require_keys, engine_id, cmdname);
+				require_keys, engine_id, cmdname, ext_sign);
 		}
 		if (ret)
 			return ret;
@@ -1095,7 +1110,7 @@ static int fit_certificate_get_data(void
 
 static int fit_certificate_add_verification_data(const char *keydir,
 		void *fit, int require_keys, int sig_noffset, const char* sign_key_name, const char *sign_key_algo,
-		const char *comment, const char *cmdname)
+		const char *comment, const char *cmdname, struct ext_sign_info *ext_sign)
 {
 	const char *node_name;
 	struct image_region *region;
@@ -1112,13 +1127,15 @@ static int fit_certificate_add_verificat
 				&region_count, &region_prop, &region_proplen))
 		return -1;
 
+	memset(&info, 0, sizeof(struct image_sign_info));
 	info.keyname = sign_key_name;
 	info.name = sign_key_algo;
 	info.crypto = image_get_crypto_algo(sign_key_algo);
 	info.keydir = keydir;
 	info.require_keys = require_keys ? "conf" : NULL;
 	info.node_offset = sig_noffset;
-	info.checksum = image_get_checksum_algo(info.name);
+	info.checksum = image_get_checksum_algo(sign_key_algo);
+	info.ext_sign = ext_sign;
 
 	if (!info.checksum || !info.crypto) {
 		printf("Unsupported signature algorithm (%s) for '%s'\n",
@@ -1128,6 +1145,7 @@ static int fit_certificate_add_verificat
 
 	ret = info.crypto->sign(&info, region, region_count, &value,
 				&value_len);
+
 	free(region);
 	if (ret) {
 		printf("Failed to sign '%s' signature node fit certificate signture node\n", node_name);
@@ -1148,7 +1166,8 @@ static int fit_certificate_add_verificat
 	return 0;
 }
 
-int fit_add_certificate_data(void *fit, const char* keydir, int require_keys, const char *sign_key_name, const char *sign_key_algo)
+int fit_add_certificate_data(void *fit, const char* keydir, int require_keys,
+	const char *sign_key_name, const char *sign_key_algo, struct ext_sign_info *ext_sign)
 {
 	int ret;
 	int sig_noffset, noffset, conf_noffset;
@@ -1159,6 +1178,11 @@ int fit_add_certificate_data(void *fit,
 		key-name-hint should be set, that certificate is what we're after. */
 	conf_noffset = fdt_path_offset(fit, FIT_CONFS_PATH);
 
+	if (ext_sign) {
+		snprintf(ext_sign->data_file, sizeof(ext_sign->data_file), "cert_%s", DATAFILE_NAME);
+		snprintf(ext_sign->sig_file, sizeof(ext_sign->sig_file), "cert_%s", SIGFILE_NAME);
+	}
+
 	if (conf_noffset < 0) {
 		debug("Can't find configuration node\n");
 		return conf_noffset;
@@ -1190,6 +1214,7 @@ int fit_add_certificate_data(void *fit,
 		info.keydir = keydir;
 		info.require_keys = require_keys ? "conf" : NULL;
 		info.crypto = image_get_crypto_algo(algo_name);
+		info.ext_sign = ext_sign;
 
 		if (!info.crypto) {
 			printf("Unsupported signature algorithm (%s) for '%s'\n",
@@ -1233,21 +1258,22 @@ int fit_add_certificate_data(void *fit,
 int fit_add_verification_data(const char *keydir, void *keydest, void *fit,
 			      const char *comment, int require_keys,
 			      const char *engine_id, const char *cmdname,
-				  const char *cert_key, const char *cert_algo)
+				  const char *cert_key, const char *cert_algo,
+				  struct ext_sign_info *ext_sign)
 {
 	int images_noffset, confs_noffset, sig_noffset;
 	int noffset;
 	int ret;
 
-	sig_noffset = fdt_subnode_offset(fit, 0, FIT_SIG_NODENAME);
-	if (sig_noffset < 0) {
-		debug("Can't find certificate signature node.\n");
-		return -FDT_ERR_NOTFOUND;
-	}
-
 	if (cert_key) {
+		sig_noffset = fdt_subnode_offset(fit, 0, FIT_SIG_NODENAME);
+		if (sig_noffset < 0) {
+			printf("Can't find certificate signature node.\n");
+			return -FDT_ERR_NOTFOUND;
+		}
+
 		ret = fit_certificate_add_verification_data(keydir, fit, require_keys,
-			sig_noffset, cert_key, cert_algo, comment, cmdname);
+			sig_noffset, cert_key, cert_algo, comment, cmdname, ext_sign);
 		if (ret)
 			return ret;
 	}
@@ -1270,7 +1296,7 @@ int fit_add_verification_data(const char
 		 */
 		ret = fit_image_add_verification_data(keydir, keydest,
 				fit, noffset, comment, require_keys, engine_id,
-				cmdname);
+				cmdname, ext_sign);
 		if (ret)
 			return ret;
 	}
@@ -1294,7 +1320,7 @@ int fit_add_verification_data(const char
 		ret = fit_config_add_verification_data(keydir, keydest,
 						       fit, noffset, comment,
 						       require_keys,
-						       engine_id, cmdname);
+						       engine_id, cmdname, ext_sign);
 		if (ret)
 			return ret;
 	}
Index: u-boot-2021.01/tools/imagetool.h
===================================================================
--- u-boot-2021.01.orig/tools/imagetool.h
+++ u-boot-2021.01/tools/imagetool.h
@@ -85,6 +85,7 @@ struct image_tool_params {
 	int bl_len;		/* Block length in byte for external data */
 	const char *engine_id;	/* Engine to use for signing */
 	bool reset_timestamp;	/* Reset the timestamp on an existing image */
+	const char *sign_tool; /* External tool for FIT signatures */
 };
 
 /*
Index: u-boot-2021.01/tools/mkimage.c
===================================================================
--- u-boot-2021.01.orig/tools/mkimage.c
+++ u-boot-2021.01/tools/mkimage.c
@@ -12,7 +12,9 @@
 #include "imximage.h"
 #include <image.h>
 #include <version.h>
-
+#include <stdio.h>
+#include <unistd.h>
+#include <limits.h>
 static void copy_file(int, const char *, int);
 
 /* parameters initialized by core will be used by the image type code */
@@ -25,6 +27,8 @@ static struct image_tool_params params =
 	.dtc = MKIMAGE_DEFAULT_DTC_OPTIONS,
 	.imagename = "",
 	.imagename2 = "",
+	.sign_tool = NULL,
+	.keyname = NULL,
 };
 
 static enum ih_category cur_category;
@@ -116,7 +120,8 @@ static void usage(const char *msg)
 		"          -r => mark keys used as 'required' in dtb\n"
 		"          -N => openssl engine to use for signing\n"
 		"          -I => include FIT certificate in image and use provided key to sign FIT certificate\n"
-		"          -L => key algorithm used to sign FIT certificate (-I)\n");
+		"          -L => key algorithm used to sign FIT certificate (-I)\n"
+		"          -S => external signing tool to use (ignores existing sign backend)\n");
 #else
 	fprintf(stderr,
 		"Signing / verified boot not supported (CONFIG_FIT_SIGNATURE undefined)\n");
@@ -154,7 +159,7 @@ static void process_args(int argc, char
 	int opt;
 
 	while ((opt = getopt(argc, argv,
-		   "a:A:b:B:c:C:d:D:e:Ef:Fk:i:K:lM:n:N:p:O:rR:qstT:vVx:I:L:")) != -1) {
+		   "a:A:b:B:c:C:d:D:e:Ef:Fk:i:K:lM:n:N:p:O:rR:qstT:vVx:I:L:S:")) != -1) {
 		switch (opt) {
 		case 'a':
 			params.addr = strtoull(optarg, &ptr, 16);
@@ -315,6 +320,9 @@ static void process_args(int argc, char
 		case 'I':
 			params.keyname = optarg;
 			break;
+		case 'S':
+			params.sign_tool = optarg;
+			break;
 		default:
 			usage("Invalid option");
 		}
Index: u-boot-2021.01/include/image.h
===================================================================
--- u-boot-2021.01.orig/include/image.h
+++ u-boot-2021.01/include/image.h
@@ -63,6 +63,7 @@ struct fdt_region;
 #include <hash.h>
 #include <linux/libfdt.h>
 #include <fdt_support.h>
+#include <ext_sign.h>
 # ifdef CONFIG_SPL_BUILD
 #  ifdef CONFIG_SPL_CRC32_SUPPORT
 #   define IMAGE_ENABLE_CRC32	1
@@ -1132,7 +1133,8 @@ int fit_cipher_data(const char *keydir,
 int fit_add_verification_data(const char *keydir, void *keydest, void *fit,
 			      const char *comment, int require_keys,
 			      const char *engine_id, const char *cmdname,
-				  const char *cert_key, const char *cert_algo);
+				  const char *cert_key, const char *cert_algo,
+				  struct ext_sign_info *ext_sign_info);
 int fit_image_verify_with_data(const void *fit, int image_noffset,
 			       const void *data, size_t size);
 int fit_image_verify(const void *fit, int noffset);
@@ -1252,9 +1254,11 @@ struct image_sign_info {
 	 */
 	const void *key;		/* Pointer to public key in DER */
 	int keylen;			/* Length of public key */
+	struct ext_sign_info *ext_sign;
 };
 
-int fit_add_certificate_data(void *fit, const char* keydir, int require_keys, const char *sign_key_name, const char *sign_key_algo);
+int fit_add_certificate_data(void *fit, const char* keydir, int require_keys,
+	const char *sign_key_name, const char *sign_key_algo, struct ext_sign_info *ext_sign);
 
 /* A part of an image, used for hashing */
 struct image_region {
Index: u-boot-2021.01/tools/Makefile
===================================================================
--- u-boot-2021.01.orig/tools/Makefile
+++ u-boot-2021.01/tools/Makefile
@@ -118,7 +118,8 @@ dumpimage-mkimage-objs := aisimage.o \
 			gpimage-common.o \
 			mtk_image.o \
 			$(RSA_OBJS-y) \
-			$(AES_OBJS-y)
+			$(AES_OBJS-y) \
+			ext_sign.o
 
 dumpimage-objs := $(dumpimage-mkimage-objs) dumpimage.o
 mkimage-objs   := $(dumpimage-mkimage-objs) mkimage.o
Index: u-boot-2021.01/tools/ext_sign.c
===================================================================
--- /dev/null
+++ u-boot-2021.01/tools/ext_sign.c
@@ -0,0 +1,126 @@
+#include <sys/stat.h>
+#include <sys/wait.h>
+#include <string.h>
+#include <unistd.h>
+#include <stdio.h>
+#include "ext_sign.h"
+
+static int write_regions(const char *path, const struct image_region region[], int region_count)
+{
+    struct stat s;
+    FILE *fdata = NULL;
+    //remove old file if it exists
+    if (!stat(path, &s) && unlink(path))
+        goto err;
+
+    fdata = fopen(path, "a+");
+    if (!fdata)
+        goto err;
+
+    for (int i = 0; i < region_count; i ++)
+        if (fwrite(region[i].data, region[i].size, 1, fdata) != 1)
+            goto err;
+
+    fclose(fdata);
+    return 0;
+err:
+    if (fdata)
+        fclose(fdata);
+    printf("Error writing data to sign\n");
+
+    return 1;
+}
+
+static uint8_t *read_signature(const char *path, size_t sig_len)
+{
+    uint8_t *sig = malloc(sig_len);
+    FILE *fsig = fopen(path, "rb");
+
+    if (!fsig)
+        goto err;
+
+    if (fread(sig, sig_len, 1, fsig) != 1)
+        goto err;
+
+    fclose(fsig);
+    return sig;
+err:
+    if (fsig)
+        fclose(fsig);
+    free(sig);
+    printf("Failed to read signature file.\n");
+
+    return NULL;
+}
+
+int ext_sign_work(struct image_sign_info *info,
+		    const struct image_region region[],
+		    int region_count, uint8_t **sigp, uint *sig_len)
+{
+    struct ext_sign_info *ext = info->ext_sign;
+    struct stat s;
+    char data_path[PATH_MAX], sig_path[PATH_MAX];
+    char key_id[128];
+    pid_t pid;
+    int status;
+    char *err_msg = NULL;
+
+    snprintf(data_path, PATH_MAX, "%s/%s", ext->work_dir, ext->data_file);
+    snprintf(sig_path, PATH_MAX, "%s/%s", ext->work_dir, ext->sig_file);
+    snprintf(key_id, sizeof(key_id), "%s?key=%s", SIGNER_ENDPOINT, info->keyname);
+
+    if (stat(ext->signer_path, &s) < 0) {
+        err_msg = "Can't find provided signer tool";
+        goto err;
+    }
+
+    if (stat(ext->work_dir, &s) < 0)
+        goto err;
+
+    if (!S_ISDIR(s.st_mode))
+        goto err;
+
+    if (write_regions(data_path, region, region_count))
+        goto err;
+
+    pid = fork();
+    if (!pid) {
+        if (execv(ext->signer_path,
+                (char *[]) {ext->signer_path, key_id, data_path, sig_path, NULL})) {
+            err_msg = "Failed to execute external signer tool.";
+            goto err;
+        }
+    } else if (pid < 0)
+        goto err;
+
+    do {
+        if (waitpid(pid, &status, 0) < 0)
+            goto err;
+    } while (!WIFEXITED(status));
+
+    printf("Signer tool (PID %d) exit: %d\n", pid, WEXITSTATUS(status));
+
+    //non-zero return = error
+    if (WEXITSTATUS(status))
+        goto err;
+
+    *sigp = read_signature(sig_path, info->crypto->key_len);
+    if (!(*sigp))
+        goto err;
+
+    *sig_len = info->crypto->key_len;
+
+    return 0;
+
+err:
+    if(err_msg)
+        printf("Error: %s\n", err_msg);
+
+    if (errno)
+        printf("%s: %d\n", strerror(errno), errno);
+
+    if (!pid)
+        exit(1);
+
+    return 1;
+}
Index: u-boot-2021.01/tools/ext_sign.h
===================================================================
--- /dev/null
+++ u-boot-2021.01/tools/ext_sign.h
@@ -0,0 +1,28 @@
+#ifndef EXT_SIGN_H
+#define EXT_SIGN_H
+#include <image.h>
+
+#define PATH_MAX 256
+#define FILE_MAX 30
+
+#define DATAFILE_NAME "data.bin"
+#define SIGFILE_NAME "signature.bin"
+#define SIGNER_ENDPOINT "boot"
+
+struct image_sign_info;
+struct image_region;
+
+/* External image sign info */
+struct ext_sign_info {
+	char signer_path[PATH_MAX]; /* The signer executable that generates signature */
+	char work_dir[PATH_MAX - FILE_MAX]; /* Path where data to be signed is found, and where signature file will be placed */
+	char data_file[FILE_MAX]; /* Data file in *work_dir* that will be signed, must be created by mkimage */
+	char sig_file[FILE_MAX]; /* Signature data file in *work_dir*, must be created by signer tool */
+};
+
+/* Executes signing step by calling some external tool. ext_sign of struct image_sign_info must be valid.*/
+int ext_sign_work(struct image_sign_info *info,
+		    const struct image_region *region,
+		    int region_count, uint8_t **sigp, uint *sig_len);
+
+#endif
Index: u-boot-2021.01/lib/rsa/rsa-sign.c
===================================================================
--- u-boot-2021.01.orig/lib/rsa/rsa-sign.c
+++ u-boot-2021.01/lib/rsa/rsa-sign.c
@@ -16,6 +16,7 @@
 #include <openssl/ssl.h>
 #include <openssl/evp.h>
 #include <openssl/engine.h>
+#include <ext_sign.h>
 
 #if OPENSSL_VERSION_NUMBER >= 0x10000000L
 #define HAVE_ERR_REMOVE_THREAD_STATE
@@ -558,6 +559,9 @@ int rsa_sign(struct image_sign_info *inf
 	ENGINE *e = NULL;
 	int ret;
 
+	if (info->ext_sign)
+		return ext_sign_work(info, (const struct image_region*) region, region_count, sigp, sig_len);
+
 	ret = rsa_init();
 	if (ret)
 		return ret;
