--- a/service/service.c
+++ b/service/service.c
@@ -1028,6 +1028,10 @@ static void _service_stopped(struct serv
 
 void service_event(const char *type, const char *service, const char *instance)
 {
+	LOG("Service event: type=%s, service=%s, instance=%s\n",
+		type ? type : "-",
+		service ? service : "-",
+		instance ? instance : "-");
 	ubus_event_bcast(type, "service", service, "instance", instance);
 }
 
