diff --git a/include/open62541/types.h b/include/open62541/types.h
index 82f6408..9bcef22 100644
--- a/include/open62541/types.h
+++ b/include/open62541/types.h
@@ -1280,6 +1280,7 @@ typedef struct {
     const UA_String *serverUris;
     size_t serverUrisSize;
     UA_Boolean useReversible;
+    UA_Boolean bareIrreversibleVariant;
 
     UA_Boolean prettyPrint;   /* Add newlines and spaces for legibility */
 
diff --git a/src/ua_types_encoding_json.c b/src/ua_types_encoding_json.c
index e36a323..8c4a307 100644
--- a/src/ua_types_encoding_json.c
+++ b/src/ua_types_encoding_json.c
@@ -1084,7 +1084,8 @@ ENCODE_JSON(Variant) {
     if(src->type == &UA_TYPES[UA_TYPES_VARIANT] && !isArray)
         wrapEO = true;
 
-    ret = writeJsonObjStart(ctx);
+    if(!ctx->bareIrreversibleVariant)
+        ret |= writeJsonObjStart(ctx);
 
     /* Write the type NodeId */
     if(ctx->useReversible) {
@@ -1096,7 +1097,8 @@ ENCODE_JSON(Variant) {
     }
 
     /* Write the Variant body */
-    ret |= writeJsonKey(ctx, UA_JSONKEY_BODY);
+    if(!ctx->bareIrreversibleVariant)
+        ret |= writeJsonKey(ctx, UA_JSONKEY_BODY);
 
     if(!isArray) {
         encodeScalarJsonWrapExtensionObject(ctx, src);
@@ -1118,7 +1120,9 @@ ENCODE_JSON(Variant) {
         }
     }
 
-    return ret | writeJsonObjEnd(ctx);
+    if(!ctx->bareIrreversibleVariant)
+        ret |= writeJsonObjEnd(ctx);
+    return ret;
 }
 
 /* DataValue */
@@ -1311,6 +1315,7 @@ UA_encodeJson(const void *src, const UA_DataType *type, UA_ByteString *outBuf,
         ctx.serverUris = options->serverUris;
         ctx.serverUrisSize = options->serverUrisSize;
         ctx.useReversible = options->useReversible;
+        ctx.bareIrreversibleVariant = options->bareIrreversibleVariant;
         ctx.prettyPrint = options->prettyPrint;
         ctx.unquotedKeys = options->unquotedKeys;
         ctx.stringNodeIds = options->stringNodeIds;
@@ -1364,6 +1369,7 @@ UA_calcSizeJson(const void *src, const UA_DataType *type,
         ctx.serverUris = options->serverUris;
         ctx.serverUrisSize = options->serverUrisSize;
         ctx.useReversible = options->useReversible;
+        ctx.bareIrreversibleVariant = options->bareIrreversibleVariant;
         ctx.prettyPrint = options->prettyPrint;
         ctx.unquotedKeys = options->unquotedKeys;
         ctx.stringNodeIds = options->stringNodeIds;
diff --git a/src/ua_types_encoding_json.h b/src/ua_types_encoding_json.h
index 0013e9e..f092419 100644
--- a/src/ua_types_encoding_json.h
+++ b/src/ua_types_encoding_json.h
@@ -27,6 +27,7 @@ typedef struct {
     uint16_t depth; /* How often did we en-/decoding recurse? */
     UA_Boolean commaNeeded[UA_JSON_ENCODING_MAX_RECURSION];
     UA_Boolean useReversible;
+    UA_Boolean bareIrreversibleVariant;
     UA_Boolean calcOnly; /* Only compute the length of the decoding */
 
     size_t namespacesSize;
