diff --git a/plugins/crypto/openssl/ua_openssl_create_certificate.c b/plugins/crypto/openssl/ua_openssl_create_certificate.c
index 2914a24..32fbda0 100644
--- a/plugins/crypto/openssl/ua_openssl_create_certificate.c
+++ b/plugins/crypto/openssl/ua_openssl_create_certificate.c
@@ -17,6 +17,7 @@
 
 #include <openssl/pem.h>
 #include <openssl/x509v3.h>
+#include <openssl/rsa.h>
 #include <openssl/err.h>
 
 /**
@@ -209,14 +210,14 @@ UA_CreateCertificate(const UA_Logger *logger, const UA_String *subject,
         goto cleanup;
     }
 
-    if(X509_gmtime_adj(X509_get_notBefore(x509), 0) == NULL) {
+    if(X509_gmtime_adj(X509_getm_notBefore(x509), 0) == NULL) {
         UA_LOG_ERROR(logger, UA_LOGCATEGORY_SECURECHANNEL,
                      "Create Certificate: Setting 'not before' failed.");
         errRet = UA_STATUSCODE_BADINTERNALERROR;
         goto cleanup;
     }
 
-    if(X509_time_adj_ex(X509_get_notAfter(x509), (int)expiresInDays, 0, NULL) ==
+    if(X509_time_adj_ex(X509_getm_notAfter(x509), (int)expiresInDays, 0, NULL) ==
        NULL) {
         UA_LOG_ERROR(logger, UA_LOGCATEGORY_SECURECHANNEL,
                      "Create Certificate: Setting 'not before' failed.");
diff --git a/plugins/crypto/openssl/ua_pki_openssl.c b/plugins/crypto/openssl/ua_pki_openssl.c
index 3b98785..369dee5 100644
--- a/plugins/crypto/openssl/ua_pki_openssl.c
+++ b/plugins/crypto/openssl/ua_pki_openssl.c
@@ -564,8 +564,8 @@ openSSL_verifyChain(CertContext *ctx, STACK_OF(X509) *stack, X509 **old_issuers,
         return UA_STATUSCODE_BADCERTIFICATECHAININCOMPLETE;
 
     /* Verification Step: Validity Period */
-    ASN1_TIME *notBefore = X509_get_notBefore(cert);
-    ASN1_TIME *notAfter = X509_get_notAfter(cert);
+    ASN1_TIME *notBefore = X509_getm_notBefore(cert);
+    ASN1_TIME *notAfter = X509_getm_notAfter(cert);
     if(X509_cmp_current_time(notBefore) != -1 || X509_cmp_current_time(notAfter) != 1)
         return (depth == 0) ? UA_STATUSCODE_BADCERTIFICATETIMEINVALID :
             UA_STATUSCODE_BADCERTIFICATEISSUERTIMEINVALID;
@@ -814,7 +814,7 @@ UA_GetCertificate_ExpirationDate(UA_DateTime *expiryDateTime,
     }
 
     /* Get the certificate Expiry date */
-    ASN1_TIME *not_after = X509_get_notAfter(x509);
+    ASN1_TIME *not_after = X509_getm_notAfter(x509);
 
     struct tm dtTime;
     ASN1_TIME_to_tm(not_after, &dtTime);
