--- a/drivers/of/of_net.c
+++ b/drivers/of/of_net.c
@@ -88,6 +88,7 @@ static void *of_get_mac_address_mtd(stru
 	u8 mac[ETH_ALEN];
 	void *addr;
 	u32 inc_idx;
+	u32 mac_val;
 
 	list = of_get_property(np, "mtd-mac-address", &size);
 	if (!list || (size != (2 * sizeof(*list))))
@@ -116,8 +117,16 @@ static void *of_get_mac_address_mtd(stru
 	if (inc_idx > 5)
 		return NULL;
 
-	if (!of_property_read_u32(np, "mtd-mac-address-increment", &mac_inc))
-		mac[inc_idx] += mac_inc;
+	if (!of_property_read_u32(np, "mtd-mac-address-increment", &mac_inc)) {
+		/* Convert to a contiguous value */
+		mac_val = (mac[3] << 16) + (mac[4] << 8) + mac[5];
+		mac_val += mac_inc << 8 * (5-inc_idx);
+
+		/* Apply the incremented value handling overflow case */
+		mac[3] = (mac_val >> 16) & 0xff;
+		mac[4] = (mac_val >> 8) & 0xff;
+		mac[5] = (mac_val >> 0) & 0xff;
+	}
 
 	if (!is_valid_ether_addr(mac))
 		return NULL;
